### PR TITLE
Remove duplicate functions from sandbox modules

### PR DESCRIPTION
**Pull request by jbrockmendel**
import canonical versions instead.

This shouldn't change any logic anywhere.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
`sandbox.utils_old.ECDF` is very similar but not identical to `distributions.empirical_distributions.ECDF`.  Is there still a use case for the utils_old version?

`sandbox.km_class` defines only one func/class, `KAPLAN_MEIER`.  Its `__init__` raises a RuntimeError telling the user to use `sandbox.survival2.KaplanMeier` instead.  GH says this was last changed 3 years ago.

`sandbox.survival` looks like it never got off the ground, was last changed 5 years ago.

docstring to `sandbox.mle` suggests it is of unknown provenance.  It's intended use is not clear.

It looks like some of these modules are pulling down reported test coverage, and others are not being imported at all, so are not counted at all in coverage results.

**Comment by jbrockmendel** [comment]
`sandbox.regression.predstd` has a `atleast_2dcol` function that is not quite identical to the `atleast_2dcols` elsewhere (this PR moves it to `sm.tools.shaping`).  The predstd version includes in its docstring "not tested because not used".

Any objection to deleting this?

**Comment by josef-pkt** [comment]
> The predstd version includes in its docstring "not tested because not used".

I'm not sure it's not a bug that it is not used, wls_prediction_std uses
`exog = np.atleast_2d(exog)`
which might be a bug if exog is 1-D. (my guess would be that there is no unit test for 1-D)

**Comment by jbrockmendel** [comment]
My thought was that the version here is close enough to the `atleast_2dcols` elsewhere that we can use that instead and delete this.

The docstring says that in a 1-dimensional case, it is interpreted as a single observation, i.e. a row vector.  `np.atleast_2d` would return a 1xN array, whereas `atleast_2dcol` would return a Nx1 array.  Sounds like the `np` version is used correctly.

**Comment by josef-pkt** [comment]
> Sounds like the np version is used correctly.

I think you are right. I was thinking of 1-D column case when we only have one column in exog, but this is for the 1-D row case.

So deleting atleast_2dcols is good.

**Comment by jbrockmendel** [comment]
sandbox.regression.runmnl and sandbox.regression.treewalkerclass look pretty similar.  Is one of them a replacement for the other?

**Comment by jbrockmendel** [comment]
Is sandbox.formula subsumed by either sm.formula or patsy?  It looks like it was ported from scikits.statsmodels in 2012 and has seen mainly compat changes since then.

**Comment by josef-pkt** [comment]
> sandbox.regression.runmnl and sandbox.regression.treewalkerclass

we don't touch those. It's very tricky to get a general tree structure for arbitrary nested discrete choice models. It would take me too long now to figure out what the status is (and to remind myself what I did)
Essentially, the related discrete choice model from a GSOC are waiting in a branch, but AFAIR, there was not much for the nested model. The intention was that those experimental modules are used for some of the underlying structure.
There are some urban economists in San Francisco and Berkeley that work on discrete choice models, and we wait for the arrival of the relevant models in python before deleting the experimental code.
(AFAIR, the treewalker class was the prototype for the datastructure in a recursive function. runmnl is a discrete choice model that is an application of it to nested multinomial logit that imposes parametric restrictions so that the discrete choices are consistent with utility maximization.)

about formula:
The old stats.models formulas are essentially dead. The implementation is pretty different from what patsy has. There are some pieces by Jonathan Taylor in the statistics computation (also for contrasts) that I never understood with my econometrics background. For those things I'm waiting until I run into them coming from a topic that provides some explanation and intuition for me.  
I doubt that keeping the old formulas provides any benefits, but I haven't looked at it in a long time.

Related: There is the old GAM code in the sandbox from the stats.models time. We never used it because the underlying C code for splines was segfaulting and we removed it. However, I think we will eventually resurrect or build something like the backfitting algorithm that is there for either some GAM versions or for partial linear models.
(Nonparametric, kernel, regression is a "10 year" project. I'm looking every once in a while but I'm not yet fully convinced that it is really "useful". It's inherently slow and even more so in python. GAM with splines works well in a penalized maximum likelihood setting but eventually we need proper kernel regression.
https://groups.google.com/d/embed/msg/pystatsmodels/KVmXLAJfMxU/lUaoV51oCwAJ and we need cross-validation and bootstrap, which means no instantaneous results. )

**Comment by jbrockmendel** [comment]
sandbox.panel.sandwich_covariance_generic.aggregate_cov has a `count` variable that gets incremented inside a loop with `count *= 1`.  I expect that was intended to be `count += 1`.  `aggregate_cov` is used in almost every other function in that module.

**Comment by jbrockmendel** [comment]
gmm.py has a spec_hausman function and a spec_hausman method.  It looks like with a little tweaking the method could dispatch to the function, keeping the logic all in one place.

**Comment by josef-pkt** [comment]
`count *= 1` is pretty sure a bug
(the function is another prototype, a nobs-nobs double loop is not very usable, and AFAICS, no other package implements this, so we cannot directly get verifying unit tests. I recently ran into a case where something like that, or with a double continuous kernel would be useful.) 

Those review comments are useful, but except for obvious cases, the code changes need to be made by topic or module group. (There is a large setup cost to get familiar with a topic again.)
The comments are difficult to find if they are hiding in PR comments.
GMM round 3 and 4 #2640, previous round was #1096
(similar for GLM #3377 #2804 #2753 to get the bells and whistles, or cleanup before getting them)

**Comment by jbrockmendel** [comment]
> Those review comments are useful, but except for obvious cases, the code changes need to be made by topic or module group.

Gotcha.  This pass through the sandbox is about done, so I'll figure out which of these count as obvious and make separate small PRs for those, then the others I'll try to find appropriate SUMM issues to attach to.

> (There is a large setup cost to get familiar with a topic again.)

Very true.  In many cases its easier to start from scratch than to figure out what was intended the last time around.  I expect that's part of the origin story for e.g. `km_class`/`survival`/`survival2`.

The current pass is focused on identifying code that may be superseded by something outside the sandbox.  The next planned pass (probably a few weeks off) is to identify sandbox code used outside the sandbox and get those dependencies tested and in shape to de-sand.

In the particular case of `sandbox.panel.sandwich_covariance_generic`, the only places it shows up are in commented-out imports that have been replaced by the `stats.sandwich_covariance` version.  So I expect it is ready to be deprecated, but it is not obvious that there are function-by-function replacements in place.

> (AFAIR, the treewalker class was the prototype for the datastructure in a recursive function. runmnl is a discrete choice model that is an application of it to nested multinomial logit that imposes parametric restrictions so that the discrete choices are consistent with utility maximization.)

That actually sounds pretty cool.  Go Bears.


In its current form, this PR should not change any logic.  It adds a module `tools.shaping` which currently holds only `atleast_2dcols`.  There are a handful of other functions spread across various `tools.utils.tools` modules that are shape-themed, but that's refactoring for another day.  Just realized `shaping.atleast_2dcols` doesn't have tests.  I'll circle back to that.

**Comment by jbrockmendel** [comment]
Related #3442

**Comment by jbrockmendel** [comment]
@josef-pkt I'll need to port the discussion above to the relevant Issues, but the actual changes in this PR should be uncontroversial.

**Comment by jbrockmendel** [comment]
@josef-pkt Would there be a chance of getting traction on this if I break it into smaller pieces?  This is just a matter of removing duplicated functions.

**Comment by josef-pkt** [comment]
maybe after the prio-high and bugfixes (plus >=3 ENH PRs) "scheduled " for this week

**Comment by josef-pkt** [comment]
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-f1d8d8822484b50619e656b808386ca54a92d618)

**new alerts:**

* 1 for Syntax error

---

*Comment posted by [lgtm.com](https://lgtm.com)*

File: statsmodels/distributions/empirical_distribution.py [hunk 1]
```diff
@@ -140,7 +140,7 @@ def __init__(self, x, side='right'):
             y = np.linspace(1./nobs,1,nobs)
             super(ECDF, self).__init__(x, y, side=side, sorted=True)
         else:
-            return interp1d(x,y,drop_errors=False,fill_values=ival)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply statsmodels/distributions/empirical_distribution.py]: Yah thats what I thought, nice to get confirmation on that.  Any idea what the intended behavior is?

[the remark this answers is not in the document]
Comment (josef-pkt) [reply statsmodels/distributions/empirical_distribution.py]: My guess is that it is the usual decision problem whether an object should have a single purpose or multiple.
EDF is a discrete distribution, but sometimes we would like a "histogram distribution" in continuous space with linear interpolation in the cdf and ppf.

**Comment by jbrockmendel** [comment]
Reducing technical debt not worth bothering; closing.

File: statsmodels/sandbox/archive/tsa.py [hunk 2]
```diff
@@ -23,28 +23,13 @@
...
-
-    if demean:
-        x = x - x.mean()
+def acovf_fft(x, demean=True):
+    """
+    acovf_fft is deprecated, will be removed [...].  It will continue
+    working until then.  Users are encouraged to
+    use sm.tsa.stattools.acovf instead.
+    """
+    return acovf(x, unbiased=False, demean=demean, fft=True, missing='none')
```
Comment (bashtage) [inline statsmodels/sandbox/archive/tsa.py]: Deprecation warning?

File: statsmodels/distributions/empirical_distribution.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bashtage) [reply statsmodels/distributions/empirical_distribution.py]: Probably remove the whole if step: and then unindent that block, and delete the else.

**Pull request closed without merging**